### PR TITLE
[SYCLomatic] Add malloc, free, and temporary allocation mappings

### PR DESCRIPTION
**Pull request by mmichel11**
In this PR, we add the APIs for `malloc`, `free`, and temporary buffers (`get_temporary_allocation` and `release_temporary_allocation`). A `tagged_pointer` type is added that contains a template parameter of type `host_sys_tag` or `device_sys_tag` specifying the location of the allocation. This type is returned by `malloc` and `get_temporary_allocation`. `tagged_pointer` currently provides basic pointer functionality and will be improved in the future once we get access to Boost iterator adaptors.

In addition to passing execution policies to the memory allocation APIs, passing a system tag type is also supported in this PR. The implemented functionality is only supported for USM mode. Static asserts are hit if `DPCT_USM_LEVEL_NONE` is set.

There are two PRs that test this functionality:
- https://github.com/oneapi-src/SYCLomatic-test/pull/426
- https://github.com/oneapi-src/SYCLomatic-test/pull/427

### CODE REVIEW

**Review by tomflinda — COMMENTED** [review]
Pls add test case in https://github.com/oneapi-src/SYCLomatic-test/tree/SYCLomatic/help_function/src for new added helper APIs in this PR

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dpl_extras/memory.h]: How about use `ptrdiff_t`?

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dpl_extras/memory.h]: `m_ptr` is not initialized.

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dpl_extras/memory.h]: Should return `const T &` here?

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dpl_extras/memory.h:806]: `operator*` and `operator->` and pointer type conversion are also needed?

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dpl_extras/memory.h:723]: Is user-defined tag that deriving `sys_tag` also permitted?

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dpl_extras/memory.h]: Why we need this base class?

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (mmichel11) [reply clang/runtime/dpct-rt/include/dpl_extras/memory.h]: Yes, I have made this change.

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (mmichel11) [reply clang/runtime/dpct-rt/include/dpl_extras/memory.h]: Fixed.

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (mmichel11) [reply clang/runtime/dpct-rt/include/dpl_extras/memory.h]: I have switched to using this.

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (mmichel11) [reply clang/runtime/dpct-rt/include/dpl_extras/memory.h:806]: I have added this, so we are not relying on conversion to pointer. I also updated the test to check `operator*` and `operator->`.

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (mmichel11) [reply clang/runtime/dpct-rt/include/dpl_extras/memory.h]: This base class just contains the shared functions / members between the standard `tagged_pointer` case and the specialization for void pointers. The CRTP template parameter is currently unused. I have removed it.

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (mmichel11) [reply clang/runtime/dpct-rt/include/dpl_extras/memory.h:723]: Since we do not currently support any custom execution policies in oneDPL, I think we should not currently allow user-defined system tags. I have put a stricter check in here to ensure this.

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (danhoeflinger) [inline clang/runtime/dpct-rt/include/dpl_extras/memory.h]: We should do our best to stay away from using non-public API of oneDPL outside of oneDPL.  Here I believe there exists `dpct::is_hetero_execution_policy<>`, you can either negate this or add a similar host version to the SYCLomatic headers.

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (danhoeflinger) [inline clang/runtime/dpct-rt/include/dpl_extras/memory.h]: Probably don't need to mention boost here explicitly as that is still under discussion and an implementation detail, but rather just mention an iterator adaptor implementation

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (danhoeflinger) [inline clang/runtime/dpct-rt/include/dpl_extras/memory.h]: May be good to explicitly use `dpct::free()` here so there is no confusion.

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (danhoeflinger) [inline clang/runtime/dpct-rt/include/dpl_extras/memory.h]: Does it make sense to abstract this out to something like:
`dpct::is_host_policy_or_tag_v<PolicyOrTag>`?

If you do this, you'd have to adjust `tag::is_policy_v`  a bit as well, probably to use `oneapi::dpl::execution::is_execution_policy_v`

Similar above.

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (danhoeflinger) [inline clang/runtime/dpct-rt/include/dpl_extras/memory.h]: May be good to explicitly use `dpct::malloc()` so there is no confusion.

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (mmichel11) [reply clang/runtime/dpct-rt/include/dpl_extras/memory.h]: I have updated this to use the negation of `dpct::is_hetero_execution_policy`.

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (mmichel11) [reply clang/runtime/dpct-rt/include/dpl_extras/memory.h]: Done

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (mmichel11) [reply clang/runtime/dpct-rt/include/dpl_extras/memory.h]: Updated

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (mmichel11) [reply clang/runtime/dpct-rt/include/dpl_extras/memory.h]: I have added `dpct::is_host_policy_or_tag_v<PolicyOrTag>` and switched `tag::is_policy_v` to just use `oneapi::dpl::execution::is_execution_policy_v` on the decayed `PolicyOrType`.

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (mmichel11) [reply clang/runtime/dpct-rt/include/dpl_extras/memory.h]: Updated

**Comment by mmichel11** [comment]
@tomflinda I have added tests for the system tag internal helper APIs in: https://github.com/oneapi-src/SYCLomatic-test/pull/427/files and a test for the `malloc_base` helper API in https://github.com/oneapi-src/SYCLomatic-test/pull/426

**Review by danhoeflinger — APPROVED** [review]
LGTM, it would be good to get and approval from @ziranzha before merging.

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (ziranzha) [reply clang/runtime/dpct-rt/include/dpl_extras/memory.h:806]: Do we need some type conversion like:
```
  template<class OtherT>
  operator tagged_pointer<Tag, OtherT>();
```

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (ziranzha) [reply clang/runtime/dpct-rt/include/dpl_extras/memory.h]: The base class just provide two simple methods:
```
  operator const T *() const { return m_ptr; }
  operator T *() { return m_ptr; }
```
I think we can remove it.

Comment (mmichel11) [reply clang/runtime/dpct-rt/include/dpl_extras/memory.h]: I have made this change.

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h
Comment (mmichel11) [reply clang/runtime/dpct-rt/include/dpl_extras/memory.h:806]: I don't think this operator is necessary since we already provide a raw pointer conversion to `otherT` which can then call the `tagged_pointer<Tag, OtherT>` constructor. This functionality is tested in `onedpl_test_tagged_pointer` in the test PR.

We also don't provide this operator in similar classes (ex: device_pointer).

**Review by ziranzha — APPROVED** [review]

**Pull request merged**
